primitives/arithmetic/src/fixed.rs

```diff
@@ -0,0 +1,1311 @@
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use sp_std::{ops::{self, Add, Sub, Mul, Div}, fmt::Debug, prelude::*, convert::{TryInto, TryFrom}};
+use codec::{Encode, Decode};
+use crate::{
+	helpers_128bit::multiply_by_rational, PerThing,
+	traits::{
+		SaturatedConversion, CheckedSub, CheckedAdd, CheckedMul, CheckedDiv, CheckedNeg,
+		Bounded, Saturating, UniqueSaturatedInto, Zero,
+	},
+};
+
+#[cfg(feature = "std")]
+use serde::{de, Deserialize, Deserializer, Serialize, Serializer};
+
+pub trait FixedPointOperand: Copy + Clone + Bounded + Zero + Saturating
+	+ PartialOrd + UniqueSaturatedInto<u128> + TryFrom<u128> + CheckedNeg {}
+
+impl FixedPointOperand for i128 {}
+impl FixedPointOperand for u128 {}
+impl FixedPointOperand for i64 {}
+impl FixedPointOperand for u64 {}
+impl FixedPointOperand for i32 {}
+impl FixedPointOperand for u32 {}
+impl FixedPointOperand for i16 {}
+impl FixedPointOperand for u16 {}
+impl FixedPointOperand for i8 {}
+impl FixedPointOperand for u8 {}
+
+pub trait FixedPointNumber:
+	Sized + Copy + Default + Debug
+	+ Saturating + Bounded
+	+ Eq + PartialEq + Ord + PartialOrd
+	+ CheckedSub + CheckedAdd + CheckedMul + CheckedDiv
+	+ Add + Sub + Div + Mul
+{
+	/// The underlying data type used for this fixed point number.
+	type Inner: Copy + Debug;
+
+	/// The accuracy of this fixed point number.
+	const DIV: Self::Inner;
+
+	/// Accuracy of this `Fixed` implementation.
+	fn accuracy() -> Self::Inner {
+		Self::DIV
+	}
+
+	/// Raw constructor. Assigns `int / Self::accuracy()` to `self`.
+	fn from_inner(int: Self::Inner) -> Self;
+
+	/// Consumes `self` and returns the inner raw value (`self * Self::accuracy()`).
+	fn into_inner(self) -> Self::Inner;
+
+	/// Creates self from an integer number `int`.
+	/// 
+	/// Returns `Self::max` or `Self::min` if `int` exceeds accuracy.
+	fn from_integer<N: UniqueSaturatedInto<Self::Inner>>(int: N) -> Self;
+
+	/// Creates `self` from an integer number `int`.
+	/// 
+	/// Returns `None` if `int` exceeds accuracy.
+	fn checked_from_integer(int: Self::Inner) -> Option<Self>;
+
+	/// Creates `self` from a rational number. Equal to `n / d`.
+	///
+	/// Panics if `d = 0`. Returns `Self::max` or `Self::min` if `n / d` exceeds accuracy.
+	fn from_rational<N: FixedPointOperand, D: FixedPointOperand>(n: N, d: D) -> Self;
```

Comment:
good target for fuzzing. It is also a great idea to write a fuzzer for this since it forces you to think about the bounds of each function and when/how they are accurate to write assertions. 

need tests to show the bounds as well.